File: golang/errno_coverage_test.go
package flyql

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type errnoCoverageExpected struct {
	Errno           int    `json:"errno,omitempty"`
	ErrnoOptions    []int  `json:"errno_options,omitempty"`
	MessageContains string `json:"message_contains,omitempty"`
}

type errnoCoverageInputConstruction struct {
	Type  string `json:"type"`
	Depth int    `json:"depth,omitempty"`
}

type errnoCoverageTest struct {
	Name              string                          `json:"name"`
	Input             string                          `json:"input,omitempty"`
	InputConstruction *errnoCoverageInputConstruction `json:"input_construction,omitempty"`
	ExpectedResult    string                          `json:"expected_result"`
	ExpectedError     errnoCoverageExpected           `json:"expected_error"`
}

type errnoCoverageFile struct {
	Version               string              `json:"version"`
	Description           string              `json:"description"`
	TestSuite             string              `json:"test_suite"`
	KnownUnreachableCodes []string            `json:"known_unreachable_codes,omitempty"`
	Tests                 []errnoCoverageTest `json:"tests"`
}

type coverageRegistryEntry struct {
	Name string `json:"name"`
}

type coverageRegistryCategory struct {
	Errors map[string]coverageRegistryEntry `json:"errors"`
}

type coverageRegistryFile struct {
	Categories map[string]coverageRegistryCategory `json:"categories"`
}

func resolveCoverageInput(t *testing.T, entry errnoCoverageTest) string {
	t.Helper()
	if entry.Input != "" || entry.InputConstruction == nil {
		return entry.Input
	}
	c := entry.InputConstruction
	switch c.Type {
	case "nested_parens":
		return strings.Repeat("(", c.Depth) + "a=1" + strings.Repeat(")", c.Depth)
	default:
		t.Fatalf("unknown input_construction type %q for %q", c.Type, entry.Name)
		return ""
	}
}

func loadErrnoCoverageFile(t *testing.T, path string) errnoCoverageFile {
	t.Helper()
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var f errnoCoverageFile
	if err := json.Unmarshal(data, &f); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
	return f
}

func loadCoverageRegistry(t *testing.T) coverageRegistryFile {
	t.Helper()
	path := filepath.Join(getTestDataDir(), "..", "..", "errors", "registry.json")
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read %s: %v", path, err)
	}
	var r coverageRegistryFile
	if err := json.Unmarshal(data, &r); err != nil {
		t.Fatalf("parse %s: %v", path, err)
	}
	return r
}

func TestCoreErrnoCoverageFixture(t *testing.T) {
	path := filepath.Join(getTestDataDir(), "parser", "errno_coverage.json")
	tf := loadErrnoCoverageFile(t, path)
	for _, tc := range tf.Tests {
		t.Run(tc.Name, func(t *testing.T) {
			input := resolveCoverageInput(t, tc)
			parser := NewParser()
			err := parser.Parse(input)
			if err == nil {
				t.Fatalf("expected error, got nil")
			}
			parseErr, ok := err.(*ParseError)
			if !ok {
				t.Fatalf("expected *ParseError, got %T", err)
			}
			if tc.ExpectedError.Errno != 0 && parseErr.Code != tc.ExpectedError.Errno {
				t.Errorf("errno: got %d, want %d (msg=%q)",
					parseErr.Code, tc.ExpectedError.Errno, parseErr.Message)
			}
			if len(tc.ExpectedError.ErrnoOptions) > 0 {
				ok := false
				for _, e := range tc.ExpectedError.ErrnoOptions {
					if parseErr.Code == e {
						ok = true
						break
					}
				}
				if !ok {
					t.Errorf("errno %d not in options %v (msg=%q)",
						parseErr.Code, tc.ExpectedError.ErrnoOptions, parseErr.Message)
				}
			}
			if tc.ExpectedError.MessageContains != "" &&
				!strings.Contains(parseErr.Message, tc.ExpectedError.MessageContains) {
				t.Errorf("message %q missing substring %q",
					parseErr.Message, tc.ExpectedError.MessageContains)
			}
		})
	}
}

func TestCoreRegistryNamesAllCovered(t *testing.T) {
	path := filepath.Join(getTestDataDir(), "parser", "errno_coverage.json")
	tf := loadErrnoCoverageFile(t, path)
	fixtureNames := make(map[string]struct{}, len(tf.Tests))
	for _, tc := range tf.Tests {
		fixtureNames[tc.Name] = struct{}{}
	}
	unreachable := make(map[string]struct{}, len(tf.KnownUnreachableCodes))
	for _, n := range tf.KnownUnreachableCodes {
		unreachable[n] = struct{}{}
	}
	reg := loadCoverageRegistry(t)
	registry := make(map[string]struct{})
	for _, e := range reg.Categories["core_parser"].Errors {
		registry[e.Name] = struct{}{}
	}
	for name := range registry {
		if _, inFixture := fixtureNames[name]; inFixture {
			continue
		}
		if _, inUnreach := unreachable[name]; inUnreach {
			continue
		}
		t.Errorf("core_parser registry code %q missing from fixture and known_unreachable_codes", name)
	}
	for name := range unreachable {
		if _, inRegistry := registry[name]; !inRegistry {
			t.Errorf("known_unreachable_codes entry %q not present in registry", name)
		}
	}
}

File: javascript/packages/flyql/test/errno-coverage.test.js
import { describe, it, expect } from 'vitest'
import fs from 'fs'
import path from 'path'
import { fileURLToPath } from 'url'
import { Parser as CoreParser, ParserError as CoreParserError } from '../src/index.js'
import { parse as columnsParse, ParserError as ColumnsParserError } from '../src/columns/index.js'

const __filename = fileURLToPath(import.meta.url)
const __dirname = path.dirname(__filename)
const REPO_ROOT = path.join(__dirname, '..', '..', '..', '..')
const REGISTRY_PATH = path.join(REPO_ROOT, 'errors', 'registry.json')
const CORE_FIXTURE_PATH = path.join(REPO_ROOT, 'tests-data', 'core', 'parser', 'errno_coverage.json')
const COLUMNS_FIXTURE_PATH = path.join(REPO_ROOT, 'tests-data', 'core', 'parser', 'columns_errno_coverage.json')

function loadJson(p) {
    return JSON.parse(fs.readFileSync(p, 'utf-8'))
}

function resolveInput(entry) {
    if ('input' in entry) {
        return entry.input
    }
    const c = entry.input_construction
    if (c.type === 'nested_parens') {
        return '('.repeat(c.depth) + 'a=1' + ')'.repeat(c.depth)
    }
    throw new Error(`unknown input_construction type: ${c.type}`)
}

function registryNames(category) {
    const reg = loadJson(REGISTRY_PATH)
    return Object.values(reg.categories[category].errors).map((e) => e.name)
}

const CORE_FIXTURE = loadJson(CORE_FIXTURE_PATH)
const COLUMNS_FIXTURE = loadJson(COLUMNS_FIXTURE_PATH)

describe('Core errno coverage fixture', () => {
    for (const entry of CORE_FIXTURE.tests) {
        it(`triggers ${entry.name}`, () => {
            const parser = new CoreParser()
            const input = resolveInput(entry)
            expect(() => parser.parse(input)).toThrow(CoreParserError)
            let thrown
            try {
                parser.parse(input)
            } catch (err) {
                thrown = err
            }
            const expected = entry.expected_error
            if ('errno' in expected) {
                expect(thrown.errno).toBe(expected.errno)
            }
            if ('errno_options' in expected) {
                expect(expected.errno_options).toContain(thrown.errno)
            }
            const mc = expected.message_contains || ''
            if (mc) {
                expect(parser.errorText).toContain(mc)
            }
        })
    }
})

describe('Columns errno coverage fixture', () => {
    for (const entry of COLUMNS_FIXTURE.tests) {
        it(`triggers ${entry.name}`, () => {
            const caps = entry.capabilities || {}
            const input = resolveInput(entry)
            let thrown
            try {
                columnsParse(input, caps)
            } catch (err) {
                thrown = err
            }
            expect(thrown).toBeInstanceOf(ColumnsParserError)
            const expected = entry.expected_error
            if ('errno' in expected) {
                expect(thrown.errno).toBe(expected.errno)
            }
            if ('errno_options' in expected) {
                expect(expected.errno_options).toContain(thrown.errno)
            }
            const mc = expected.message_contains || ''
            if (mc) {
                expect(thrown.message).toContain(mc)
            }
        })
    }
})

describe('Registry name coverage', () => {
    it('core_parser registry names are all covered (fixture or known-unreachable)', () => {
        const fixtureNames = new Set(CORE_FIXTURE.tests.map((t) => t.name))
        const unreachable = new Set(CORE_FIXTURE.known_unreachable_codes || [])
        const registry = new Set(registryNames('core_parser'))
        const missing = [...registry].filter((n) => !fixtureNames.has(n) && !unreachable.has(n))
        expect(missing).toEqual([])
        const unknownUnreachable = [...unreachable].filter((n) => !registry.has(n))
        expect(unknownUnreachable).toEqual([])
    })

    it('columns_parser registry names are all covered (fixture or known-unreachable)', () => {
        const fixtureNames = new Set(COLUMNS_FIXTURE.tests.map((t) => t.name))
        const unreachable = new Set(COLUMNS_FIXTURE.known_unreachable_codes || [])
        const registry = new Set(registryNames('columns_parser'))
        const missing = [...registry].filter((n) => !fixtureNames.has(n) && !unreachable.has(n))
        expect(missing).toEqual([])
        const unknownUnreachable = [...unreachable].filter((n) => !registry.has(n))
        expect(unknownUnreachable).toEqual([])
    })
})
